src/nvim/lua/treesitter.c

```diff
@@ -782,3 +784,63 @@ static int query_tostring(lua_State *L)
   lua_pushstring(L, "<query>");
   return 1;
 }
+
+static int query_inspect(lua_State *L)
+{
+  TSQuery *query = query_check(L, 1);
+  if (!query) {
+    return 0;
+  }
+
+  uint32_t n_pat = ts_query_pattern_count(query);
+  lua_createtable(L, 0, 2); // [retval]
+  lua_createtable(L, n_pat, 1); // [retval, patterns]
+  for (size_t i = 0; i < n_pat; i++) {
+    uint32_t len;
+    const TSQueryPredicateStep *step = ts_query_predicates_for_pattern(query,
+                                                                       i, &len);
+    if (len == 0) {
+      continue;
+    }
+    lua_createtable(L, len/4, 1);  // [retval, patterns, pat]
+    lua_createtable(L, 3, 0);  // [retval, patterns, pat, pred]
+    int nextpred = 1;
+    int nextitem = 1;
+    for (size_t k = 0; k < len; k++) {
+      if (step[k].type == TSQueryPredicateStepTypeDone) {
```

Comment:
Perhaps, but I will consider matter of style after the complete implementation is done. It might turn out lua is to slow for this, then we will do predicate matching in C instead.